rfc/0000-batch-cancellation.rst

```diff
@@ -0,0 +1,538 @@
+===========================
+Batch Cancellation Strategy
+===========================
+
+.. author:: Jackie Goldstein
+.. date-accepted:: Leave blank. This will be filled in when the proposal is accepted.
+.. ticket-url:: Leave blank. This will eventually be filled with the ticket URL which will track the progress of the implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first Hail version which implements the described feature.
+.. header:: This proposal is `discussed at this pull request <https://github.com/hail-is/hail-rfc/pull/0>`_. **After creating the pull request, edit this file again, update the number in the link, and delete this bold sentence.**
+.. sectnum::
+.. contents::
+.. role:: python(code)
+
+Hail Batch allows users to execute containerized workflows on a
+multi-tenant compute cluster. Users submit a single batch which
+consists of the set of job specifications to execute and the
+dependencies between jobs. A major part of the Batch system is
+cancellation. A cancellation event can be initiated by the user or by
+the system and causes all currently running jobs in the batch to be
+terminated and no further jobs can be executed (unless they are set to
+always_run). This feature is important for stopping workflows that are
+not behaving as expected and for controlling costs. For example, a
+user might want to cancel a batch if they notice all of their jobs are
+failing or their jobs are taking substantially longer than
+expected. The system also cancels batches if the billing limit for the
+project has been exceeded or the user has specified they want their
+batch to be automatically cancelled after N failures. The current
+implementation of cancellation in Batch relies on a series of tables
+that keep track of the number of jobs that are **cancellable** with
+different job states such as running, ready, and creating. When a user
+cancels a batch, a single database operation instantly changes the
+workload demands on the system causing the autoscaler and scheduler to
+react immediately to the cancellation event by not spinning up
+resources for the cancelled jobs and not scheduling the jobs on
+existing workers. While this implementation is extremely performant,
+the database structures necessary to bookkeep the number of resources
+in each state is not flexible for new features we'd like to add to the
+Batch system such as job groups. It also is tricky to understand and
+modify this part of the code base leading to more developer overhead
+and slower feature development. This proposal reviews how cancellation
+works in the current system and compares options for other
+cancellation implementations before recommending a new simpler
+implementation.
+
+
+----------
+Motivation
+----------
+
+Cancellation is a key operation in the Batch system. It must be as
+fast as possible to avoid provisioning resources for jobs that users
```

Comment:
You do this later on in the constraints section, maybe good to reflect those up here.

Reply:
I'm not entirely certain what the expectations should be, but I think instead of talking about time complexity or latency, we should be talking about wasted resources and/or USD.

Reply:
That said, I think starting with talking about time-complexity is good. It's a lot easier to think about than the implications on resources and cost.